File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        best = 0
        bestStart = 0
        def expand(l,r,best,bestStart):
            while l >= 0 and r < n and s[l] == s[r]:
                l -= 1
                r += 1
            length = r-l-1
            if length > best:
                best      = length
                bestStart = l+1
            return (best,bestStart)
        for i in range(n):
            (best,bestStart) = expand(i,i,best,bestStart)
            (best,bestStart) = expand(i,i+1,best,bestStart)
        return s[bestStart:bestStart+best]

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int best = 0;
        string ans;
        int n = (int)s.size();
        auto expand = [&s,&best,&ans,n](int l, int r){
            while (l >= 0 && r < n && s[l] == s[r]){
                l--;
                r++;
            }
            int len = r-l-1;
            best = max(best,len);
            if (best == len)   
                ans = s.substr(l+1,len);
        };
        for (int i=0; i<n; i++){
            expand(i,i);
            expand(i,i+1);
        }
        return ans;
    }
};